src/core/thread/mle_router.cpp

```diff
@@ -99,14 +99,21 @@ void MleRouter::HandlePartitionChange(void)
     netif.GetCoap().AbortTransaction(&MleRouter::HandleAddressSolicitResponse, this);
 }
 
-bool MleRouter::IsRouterRoleEnabled(void) const
+bool MleRouter::IsRouterEligible(void) const
 {
-    return mRouterRoleEnabled && IsFullThreadDevice();
+    return (IsFullThreadDevice() &&
+            (GetNetif().GetKeyManager().GetSecurityPolicyFlags() & OT_SECURITY_POLICY_ROUTERS) && mRouterEligible);
 }
 
-void MleRouter::SetRouterRoleEnabled(bool aEnabled)
+otError MleRouter::SetRouterEligible(bool aEnabled)
 {
-    mRouterRoleEnabled = aEnabled;
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(IsFullThreadDevice() &&
+                     (GetNetif().GetKeyManager().GetSecurityPolicyFlags() & OT_SECURITY_POLICY_ROUTERS),
```

Comment:
I can't shake the feeling that this `VerifyOrExit` will occasionally jump over `SetBeaconEnabled` and/or `BecomeDetached` calls below. I'm thinking about the complex corner case like: device starts, `mRouterEligible` is `true` from constructor, and then somebody calls `SetRouterEligible( false )` which before would call `BecomeDetached()` and now it wont. Maybe that is a problem, maybe not (after all, device would need to become leader or router first) - though it is far from obvious to me that this code is fully correct.

Like, is it possible for Security Policy TLV to change bit after devices are commissioned? Existing leader & routers would need to drop their roles, I don't see a code like that.

Reply:
Thanks for raising this issue.  I added some logic in `Mle::HandleStateChanged()` to address this issue.  Please review.

Reply:
LGTM